### DISCUSSION TITLE
kustomize, regex versioning not detected, registryUrls ignored

### DISCUSSION THREAD

**Discussion by lukasmrtvy** [renovatebot/renovate#7950] [Request Help]
### Which Renovate are you using?

Renovate Open Source CLI

### Which platform are you using?

GitLab self-hosted

### Have you checked the logs? Don't forget to include them if relevant

```
{
    "packageFile": "foo/hello-world/kustomization.yaml",
    "manager": "kustomize",
    "deps": [
    {
        "datasource": "docker",
        "versioning": "docker",
        "depName": "example.com/foo/hello-world",
        "currentValue": "1-rolling",
        "depIndex": 0,
        "updates": [
        {
            "fromVersion": "1",
            "toVersion": "77",
            "newValue": "77-rolling",
            "newMajor": 77,
            "newMinor": null,
            "updateType": "major",
            "isSingleVersion": true,
            "newVersion": "77"
        }
        ],
        "warnings": [],
        "fixedVersion": "1-rolling",
        "dockerRegistry": "https://example.com",
        "dockerRepository": "foo/hello-world"
    }
    ]
},
```

### What would you like to do?

I am trying to bumping major version ( `<number>-rolling` ) in kustomize manager via package rule, but:
1. wrong versioning is detected  ( docker instead regex )
2. its not possible to use `registryUrls` and  `packageNames` together ( its possible to use `example.com/foo/hello-world` as packageName, but its quite redundant if packageNames > 1 )

renovate.json
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base"],
  "prHourlyLimit": 60,
  "prConcurrentLimit": 10,
  "kustomize": {
    "packageRules": [{
        "packageNames": [
          "hello-world"
        ],
        "versioning": "regex:^(?<major>\\d+)-rolling",
        "updateTypes": ["major"],
        "registryUrls": ["https://example.com/foo"],
        "enabled": true
    }]
  }
}
```
foo/hello-world/kustomization.yaml
```
resources:
- deployment.yaml

images:
- name: example.com/foo/hello-world
  newTag: 1-rolling
```

Thanks for help. I am not sure if I am hitting multiple bugs or its just wrong config. But kustomize manager is not so well tested https://github.com/renovatebot/renovate/issues/7783

**Comment by viceice** [comment]
1. You do not need a regex manager, docker major updates are  simply disabled by default, so you need to enable it (we have a preset for it)

2. `packageNames` need to match full depName, so for docker images the registry url is part of depName.

**Comment by lukasmrtvy** [comment]
I am not sure if I follow:
1. I did not mention regex manager, only regex versioning.. and I have enabled "updateTypes": ["major"] , seems that log output ^ is wrong with "versioning": "docker" line then...
2. So I have to use always:
```
"packageRules": [{
        "packageNames": [
          "myregistry.example.com/foo/hello-world-1",
          "myregistry.example.com/foo/hello-world-2",
          "myregistry.example.com/foo/hello-world-2"
        ],
```
right? No way how to group packages per registry ?

**Comment by viceice** [comment]
1. Sure, I was wrong. I had to say regex versioning. 🙈

2. Yes, but you can use `packagePatterns` regexes instead / combining with `packageNames`

**Comment by viceice** [comment]
https://docs.renovatebot.com/presets-docker/#dockerenablemajor is how to enable major updates for docker.

**Comment by viceice** [comment]
You can use the following to additionally match packages in rules.

https://docs.renovatebot.com/configuration-options/#packagepatterns

https://docs.renovatebot.com/configuration-options/#sourceurlprefixes

**Comment by viceice** [comment]
to get renovate working on kustomize you need to add `kind: Kustomization`  to your file

https://github.com/viceice-tests/kustomization/blob/f3fc0eb36a05f0786b25af5aaed5d4fe5dc2ee06/kustomization.yaml#L1

https://github.com/renovatebot/renovate/blob/3b1ed2f4656ea216192f21ec3200ec079098d8aa/lib%2Fmanager%2Fkustomize%2Fextract.ts#L74

see viceice-tests/kustomization#1 for sample

**Comment by viceice** [comment]
So i suggest the following config for you
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base", ":dependencyDashboard"],
  "prHourlyLimit": 60,
  "prConcurrentLimit": 10,
  "packageRules": [
    {
      "packagePatterns": [
        "^example\\.com\\/foo\\/hello-world"
      ],
      "groupName": "some-group"
    }
  ]
}
```

Remove the packagerout, if you don't want to group. Also the [dependency dashboard](https://docs.renovatebot.com/configuration-options/#dependencydashboard) is very helpful, sample is [here](viceice-tests/kustomization#2) and [here](renovatebot/renovate#2958)

**Comment by lukasmrtvy** [comment]
Thanks ! ❤️ 

Works for me:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base", ":dependencyDashboard"],
  "prHourlyLimit": 60,
  "prConcurrentLimit": 10,
  "packageRules": [{
      "packageNames": [
        "example.com/foo/hello-world-1",
        "example.com/foo/hello-world-2",
        "example.com/foo/hello-world-3"
      ],
      "versioning": "regex:^(?<major>\\d+)-rolling",
      "updateTypes": ["major"],
      "enabled": true
  }]
}
```

Actually I had `kind: Kustomization` in kustomization.yaml, but its tricky, there is no mention in docs ( https://docs.renovatebot.com/modules/manager/kustomize/ ) and kustomize itself will work without this piece of code.

**Comment by viceice** [comment]
1. you do not need the custom regex versioning, renovate docker versioning will to it
2. major updates are already enabled

Please use:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:base", ":dependencyDashboard"],
  "prHourlyLimit": 60,
  "prConcurrentLimit": 10
}
```

**Comment by rarkins** [comment]
I don't understand how it's working with updateTypes in that package rule

**Comment by lukasmrtvy** [comment]
Me neither, I am aware of docker versioning, but I want to bumping explicitly over `<number>-rolling` format, docker versioning will not me help here, if anyone will switch to `<number>-feature`, it should not get PRs for upgrade.

**Comment by rarkins** [comment]
You would need to do it this way then:

```
  "packageRules": [{
      "packageNames": [
        "example.com/foo/hello-world-1",
        "example.com/foo/hello-world-2",
        "example.com/foo/hello-world-3"
      ],
      "versioning": "regex:^(?<major>\\d+)-rolling"
  }, {
      "packageNames": [
        "example.com/foo/hello-world-1",
        "example.com/foo/hello-world-2",
        "example.com/foo/hello-world-3"
      ],
      "updateTypes": ["major"],
      "enabled": true
  }]
```

First rule is to apply the versioning to those package names.

Second rule is for *after* lookup and versioning has been done, to then enable major updates.

**Comment by lukasmrtvy** [comment]
^ Problem is that package example.com/foo/hello-world is detected as "manager": "kustomize", "datasource": "docker" and **"versioning": "docker"**, so regex vesioning is not applied, thats why major updates are working, its in log in OP.  Not sure whats wrong, even:
```
  "packageRules": [{
      "packageNames": [
        "example.com/foo/hello-world"
      ],
      "versioning": "regex:^(?<major>\\d+)-rolling"
  }, {
      "packageNames": [
        "example.com/foo/hello-world-1"
      ],
      "updateTypes": ["major"],
      "enabled": true
  }]
```
nor
```
  "packageRules": [{
      "packagePatterns": [
        "^example\\.com\\/foo\\/.*"
      ],
      "versioning": "regex:^(?<major>\\d+)-rolling"
  }, {
      "packagePatterns": [
         "^example\\.com\\/foo\\/.*"
      ],
      "updateTypes": ["major"],
      "enabled": true
  }]
```
rules are not working.. 

EDIT:
Seems like a bug, its not possible to use regex versioning with kustomize, can You confirm it? I had dozens of renovate.json configurations, but I am still getting:
```
            "packageFile": "foo/hello-world/kustomization.yaml",
             "manager": "kustomize",
             "deps": [
               {
                 "datasource": "docker",
                 "versioning": "docker",
                 "depName": "example.com/foo/hello-world",
                 "currentValue": "1-rolling",
                 "depIndex": 1,
                 "updates": [
                   {
                     "fromVersion": "1-rolling",
                     "toVersion": "106-rolling",
                     "newValue": "106-rolling",
                     "newMajor": 106,
                     "newMinor": 0,
                     "updateType": "major",
                     "isSingleVersion": true
                   }
                 ],
                 "warnings": [],
                 "fixedVersion": "1-rolling",
                 "dockerRegistry": "https://example.com",
                 "dockerRepository": "foo/hello-world"
               }
             ]
           },
```

**Comment by lukasmrtvy** [comment]
@rarkins its not possible to switch versioning in kustomize manager, isnt it hardcoded somehow ?

**Comment by viceice** [comment]
@lukasmrtvy can you please provide a public reproduction repo. Please also describe again what pr you expect and what pr you wrongly get.

I've provided a minimal repo, which does what you want above, so it's not clear what you are expecting.

**Comment by lukasmrtvy** [comment]
@viceice

As You can see here https://gitlab.com/lukasmrtvy/kustomization-renovate/-/jobs/894051757, its not possible to set regex versioning for docker images in kustomize manager, even with @rarkins snippet. ( .. and I dont want to use loose/docker versioning.. )

- https://gitlab.com/lukasmrtvy/kustomization-overlay
- https://gitlab.com/lukasmrtvy/kustomization-base
- https://gitlab.com/lukasmrtvy/kustomization-renovate


https://github.com/renovatebot/config-help/issues/1013 ( remote bases problem ) is solved ( workaround ) by https://github.com/renovatebot/config-help/issues/1013#issuecomment-738707195

**Comment by viceice** [comment]
Ok, not sure why versioning isn't working, will try to debug.

What is the problem to **not** using `docker` versioning? It will do what you want. It will only suggest major updates from `19-slim` docker tag, it will never update to `19` nor `19.0.1-slim`.

**Comment by lukasmrtvy** [comment]
@viceice I have to use regex for more complex use case, also I want to use maven versioning for other environemnts ( other than dev env ) and this will definetly will not work cuz of some strange bug in kustomize..

**Comment by viceice** [comment]
ok, docker versioning is assigned [here](https://github.com/renovatebot/renovate/blob/3b1ed2f4656ea216192f21ec3200ec079098d8aa/lib%2Fmanager%2Fkustomize%2Fextract.ts#L53)

I think the regex versioning will be applied, but the log will show the extracted deps **without** applied package rules, so you need to make your sample some more complex to see the diffenrences between docker and regex versioning. So maybe your regex is simply wrong and doesn't match what you expect.

**Comment by lukasmrtvy** [comment]
I switched versioning to loose https://gitlab.com/lukasmrtvy/kustomization-overlay/-/blob/6c89898a0c7fc8c87221db350ab862f56e5e576e/renovate.json#L11,
but renovate is still showing `"versioning": "docker"` ( https://gitlab.com/lukasmrtvy/kustomization-renovate/-/jobs/894584293#L283 ) on that package..

**Comment by viceice** [comment]
as you can sere here
![image](https://user-images.githubusercontent.com/1798109/101350419-33795300-388f-11eb-8d36-614afd07f5c0.png)

It is using loose versioning because docker versioning won't  suggest to update from `19-slim` to `19.239.11-slim`.
So as i said, the log shows the extracted values **without** package rule applied!



BTW: please remove [this](https://gitlab.com/lukasmrtvy/kustomization-renovate/-/blob/master/.gitlab-ci.yml#L7) line, no longer needed and can cause issues.

**Comment by lukasmrtvy** [comment]
> So as i said, the log shows the extracted values without package rule applied!

Oh, that makes debugging very hard. Would be nice to log actual versioning used for packageRule. This is really confusing, but again, I am newbie ( Iack of knowledge regarding renovate ). Thanks ! ❤️ 

P.S. I removed that line.

For example instead of this:
```
               {
                 "datasource": "docker",
                 "versioning": "docker",
                 "depName": "renovate/renovate",
                 "currentValue": "19-slim",
                 "depIndex": 1,
                 "updates": [
                   {
                     "fromVersion": "19-slim",
                     "toVersion": "19.239.11-slim",
                     "newValue": "19.239.11-slim",
                     "newMajor": 19,
                     "newMinor": 239,
                     "updateType": "minor",
                     "isSingleVersion": true
                   },
                   {
                     "fromVersion": "19-slim",
                     "toVersion": "23.96.0-slim",
                     "newValue": "23.96.0-slim",
                     "newMajor": 23,
                     "newMinor": 96,
                     "updateType": "major",
                     "isSingleVersion": true
                   }
                 ],
                 "warnings": [],
                 "fixedVersion": "19-slim",
                 "sourceUrl": "https://github.com/renovatebot/renovate",
                 "dockerRegistry": "https://index.docker.io",
                 "dockerRepository": "renovate/renovate"
               }
```

use:

```
               {
                 "datasource": "docker",
                 "defaultVersioning": "docker",
                 "depName": "renovate/renovate",
                 "currentValue": "19-slim",
                 "depIndex": 1,
                 "updates": [
                   {
                     "fromVersion": "19-slim",
                     "toVersion": "19.239.11-slim",
                     "newValue": "19.239.11-slim",
                     "newMajor": 19,
                     "newMinor": 239,
                     "updateType": "minor",
                     "isSingleVersion": true,
                     "detectedVersioning": "loose"
                   },
                   {
                     "fromVersion": "19-slim",
                     "toVersion": "23.96.0-slim",
                     "newValue": "23.96.0-slim",
                     "newMajor": 23,
                     "newMinor": 96,
                     "updateType": "major",
                     "isSingleVersion": true,
                     "detectedVersioning": "docker"
                   }
                 ],
                 "warnings": [],
                 "fixedVersion": "19-slim",
                 "sourceUrl": "https://github.com/renovatebot/renovate",
                 "dockerRegistry": "https://index.docker.io",
                 "dockerRepository": "renovate/renovate"
               }
```
